File: Other test/normal_equations/LinearCPU.cpp
#include <Eigen/Dense>
#include "matplotlibcpp.h"
#include <random>
#include <chrono>

namespace plt = matplotlibcpp;

int main() {
  // Generate random dataset
  std::default_random_engine generator{0};
  const int n_samples = 20000000;
  Eigen::VectorXd X(n_samples);
  Eigen::VectorXd y(n_samples);

  auto start_time_gen = std::chrono::high_resolution_clock::now();
  for (int i = 0; i < n_samples; ++i) {
    X(i) = i * 0.5;
    y(i) = 50 + 0.5 * X(i) + std::normal_distribution<double>(0, 200)(generator);
  }
  auto end_time_gen = std::chrono::high_resolution_clock::now();
     // Print CPU run time
  std::chrono::duration<double> gen_time = end_time_gen - start_time_gen;
  std::cout << "Generating time: " << gen_time.count() << " seconds" << std::endl;

  // Add a column of ones to X to account for the intercept term
  Eigen::MatrixXd X_ones(n_samples, 2);
  X_ones.col(0).setOnes();
  X_ones.col(1) = X;

  // Compute normal equation
  auto start_time = std::chrono::high_resolution_clock::now();
  const auto XtX = X_ones.transpose() * X_ones;
  const auto XtX_inv = XtX.inverse();
  const auto Xty = X_ones.transpose() * y;
  const auto theta = XtX_inv * Xty;
  auto end_time = std::chrono::high_resolution_clock::now();

   // Print CPU run time
  std::chrono::duration<double> cpu_time = end_time - start_time;
  std::cout << "CPU run time: " << cpu_time.count() << " seconds" << std::endl;

//   // Plot initial data points
//   std::vector<double> x_vec(X.data(), X.data() + X.size());
//   std::vector<double> y_vec(y.data(), y.data() + y.size());
// //   plt::scatter(x_vec, y_vec, 5);

//   // Plot regression line
//   const double x_min = X.minCoeff();
//   const double x_max = X.maxCoeff();
//   std::vector<double> x_vals{x_min, x_max};
//   std::vector<double> y_vals{theta(0) + theta(1) * x_min, theta(0) + theta(1) * x_max};

//   plt::plot(x_vals, y_vals, "--r");

//   // Add labels and title to plot
//   plt::xlabel("X");
//   plt::ylabel("y");
//   plt::title("Linear Regression");

//   // Show plot
//   plt::show();

  // Compute error on predicted values
  auto start_time_mse = std::chrono::high_resolution_clock::now();
  const auto y_pred = X_ones * theta;
  const auto mse = (y_pred - y).squaredNorm() / n_samples;
  auto end_time_mse = std::chrono::high_resolution_clock::now();

//   std::printf("Mean Squared Error: %.2f\n", mse);

     // Print CPU run time
  std::chrono::duration<double> mse_time = end_time_mse - start_time_mse;
  std::cout << "MSE run time: " << mse_time.count() << " seconds" << std::endl;

  return 0;
}

File: Other test/normal_equations/LinearCPU.py
import numpy as np
import matplotlib.pyplot as plt
import time

# Generate random dataset
np.random.seed(0)
n_samples = 20
X = np.linspace(0, 10, n_samples)
y = 50 + 0.5 * X + np.random.normal(0, 200, n_samples)

# Add a column of ones to X to account for the intercept term
X_ones = np.hstack([np.ones((n_samples, 1)), X.reshape(-1, 1)])

# Compute normal equation
start_time = time.time()
XtX = np.dot(X_ones.T, X_ones)
XtX_inv = np.linalg.inv(XtX)
Xty = np.dot(X_ones.T, y)
theta = np.dot(XtX_inv, Xty)
end_time = time.time()

# Print CPU run time
cpu_time = end_time - start_time
print(f"CPU run time: {cpu_time:.5f} seconds")

# Plot initial data points
plt.scatter(X, y, s=5)

# Plot regression line
x_vals = np.array([X.min(), X.max()])
y_vals = theta[0] + theta[1] * x_vals
plt.plot(x_vals, y_vals, "--r")

# Add labels and title to plot
plt.xlabel("X")
plt.ylabel("y")
plt.title("Linear Regression")

# Show plot
plt.show()

# Compute error on predicted values
y_pred = np.dot(X_ones, theta)
mse = np.mean((y_pred - y)**2)
print(f"Mean Squared Error: {mse:.2f}")
